### PR TITLE
Create witnet_reputation (TRS and ARS)

### LINKED ISSUE #597: Total Reputation Set implementation

### LINKED ISSUE #598: Active Reputation Set implementation

### PR DESCRIPTION
**Pull request by lrubiorod**
Implemented reputation system declared in https://github.com/witnet/research/issues/24
Added tests

Close #597 
Close #598

### CODE REVIEW

File: reputation/src/ars.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        self.queue.size()
+    }
+
+    /// Contains method for `ActiveReputationSet<K>`
+    pub fn contains(&self, key: &K) -> bool {
+        self.map.contains_key(key)
+    }
+
+    /// Returns an iterator of active identities
+    pub fn active_identities(&self) -> std::collections::hash_map::Keys<'_, K, u16> {
```
Comment (tmpolaczyk) [inline reputation/src/ars.rs]:
```suggestion
    pub fn active_identities(&self) -> impl Iterator<Item = &K> {
```

File: reputation/src/trs.rs [hunk 2]
```diff
@@ -330,44 +330,54 @@ where
...
+}
 
-    // Increment a cache entry
-    fn inc_cache(map: &mut HashMap<K, V, S>, k: K, v: V) {
-        let zero = V::default();
-        if v != zero {
-            *map.entry(k).or_default() += v;
-        }
+/// Increment a cache entry
+pub fn inc_cache<K, V, S>(map: &mut HashMap<K, V, S>, k: K, v: V)
```
Comment (aesedepece) [inline reputation/src/trs.rs]: No abbreviations, please :wink:

File: reputation/src/trs.rs [hunk 3]
```diff
@@ -0,0 +1,970 @@
...
+    iter,
+    ops::{AddAssign, SubAssign},
+};
+
+use failure::Fail;
+
+/// Total Reputation Set
+///
+/// This data structure keeps track of the total reputation `V`
+/// associated to every identity `K`. Reputation is issued in "coins" which
```
Comment (aesedepece) [inline reputation/src/trs.rs]: `coins` → `reputation packets`

File: reputation/src/trs.rs [hunk 4]
```diff
@@ -0,0 +1,970 @@
...
+                break;
+            }
+        }
+
+        assert!(to_subtract.is_empty(), "Mismatch between cache and queue");
+
+        Ok(total_subtracted)
+    }
+
+    fn spend_coins(rep_diff: &mut HashMap<K, V, S>, to_subtract: &mut HashMap<K, V, S>) {
```
Comment (aesedepece) [inline reputation/src/trs.rs]:
```suggestion
    fn expire_packets(rep_diff: &mut HashMap<K, V, S>, to_subtract: &mut HashMap<K, V, S>) {
```

File: reputation/src/trs.rs [hunk 5]
```diff
@@ -0,0 +1,970 @@
...
+                }
+            });
+        }
+    }
+
+    // Subtract `ts` from coin `x`. Returns (retain_x, retain_ts).
+    // if x > ts, keep coin but remove `ts`
+    // if x == ts, remove both
+    // if x < ts, remove coin but keep `ts`
+    fn spend_coin(x: &mut V, ts: &mut V) -> (bool, bool) {
```
Comment (aesedepece) [inline reputation/src/trs.rs]:
```suggestion
    fn expire_packet(x: &mut V, ts: &mut V) -> (bool, bool) {
```

File: reputation/src/ars.rs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    /// ```
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            map: HashMap::new(),
+            queue: CircularBuffer::new(capacity),
+        }
+    }
+
+    /// Gets the capacity of the `ActiveReputationSet<K>`
+    pub fn capacity(&self) -> usize {
```
Comment (tmpolaczyk) [inline reputation/src/ars.rs]: For consistency with `buffer_size()`

```suggestion
    pub fn buffer_capacity(&self) -> usize {
```

File: reputation/src/ars.rs [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    /// assert_eq!(ars.capacity(), 3);
+    /// ```
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            map: HashMap::new(),
+            queue: CircularBuffer::new(capacity),
+        }
+    }
+
+    /// Gets the capacity of the `ActiveReputationSet<K>`
```
Comment (tmpolaczyk) [inline reputation/src/ars.rs]:
```suggestion
    /// Gets the capacity of the buffer: the number of insertions that it takes to start dropping old entries.
```

File: reputation/src/ars.rs [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    pub fn contains(&self, key: &K) -> bool {
+        self.map.contains_key(key)
+    }
+
+    /// Returns an iterator of active identities
+    pub fn active_identities(&self) -> impl Iterator<Item = &K> {
+        self.map.keys()
+    }
+
+    /// Returns the length of active identities
```
Comment (tmpolaczyk) [inline reputation/src/ars.rs]:
```suggestion
    /// Returns the number of active identities
```

File: reputation/src/ars.rs [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    pub fn active_identities_len(&self) -> usize {
+        self.map.len()
+    }
+
+    /// Method to add a new entry
```
Comment (tmpolaczyk) [inline reputation/src/ars.rs]:
```
/// If the buffer is full, the oldest entry will be dropped, and the identity cache will be accordingly updated. 
```

**Pull request merged**